ruby/ql/lib/codeql/ruby/frameworks/ActionDispatch.qll

```diff
@@ -0,0 +1,889 @@
+    private ConditionalExpr e;
+
+    ConditionalRouteBlock() { this = TConditionalRouteBlock(parent, e) }
+
+    override string getAPrimaryQlClass() { result = "ConditionalRouteBlock" }
+
+    override Stmt getAStmt() { result = e.getBranch(_).(StmtSequence).getAStmt() }
+
+    override string getPathComponent() { none() }
+
+    override string getControllerComponent() { none() }
+
+    override string toString() { result = e.toString() }
+
+    override Location getLocation() { result = e.getLocation() }
+  }
+
+  /**
+   * A route block defined by a call to `namespace`.
+   * ```rb
+   * namespace :admin do
+   *   resources :posts
+   * end
+   * ```
+   * https://api.rubyonrails.org/classes/ActionDispatch/Routing/Mapper/Scoping.html#method-i-namespace
+   */
+  private class NamespaceRouteBlock extends NestedRouteBlock, TNamespaceRouteBlock {
+    private MethodCall call;
+    private Block block;
+
+    NamespaceRouteBlock() { this = TNamespaceRouteBlock(parent, call, block) }
+
+    override Stmt getAStmt() { result = block.getAStmt() }
+
+    override string getPathComponent() { result = this.getNamespace() }
+
+    override string getControllerComponent() { result = this.getNamespace() }
+
+    private string getNamespace() {
+      call.getArgument(0).getConstantValue().isStringOrSymbol(result)
+    }
+
+    override string toString() { result = call.toString() }
+
+    override Location getLocation() { result = call.getLocation() }
+  }
+
+  /**
+   * A route configuration. See `Route` for more info
+   */
+  newtype TRoute =
+    /**
+     * See `ExplicitRoute`
+     */
+    TExplicitRoute(RouteBlock b, MethodCall m) {
+      b.getAStmt() = m and m.getMethodName() = anyHttpMethod()
+    } or
+    /**
+     * See `ResourcesRoute`
+     */
+    TResourcesRoute(RouteBlock b, MethodCall m, string action) {
+      b.getAStmt() = m and
+      m.getMethodName() = "resources" and
+      action in ["show", "index", "new", "edit", "create", "update", "destroy"] and
+      applyActionFilters(m, action)
+    } or
+    /**
+     * See `SingularResourceRoute`
+     */
+    TResourceRoute(RouteBlock b, MethodCall m, string action) {
+      b.getAStmt() = m and
+      m.getMethodName() = "resource" and
+      action in ["show", "new", "edit", "create", "update", "destroy"] and
+      applyActionFilters(m, action)
+    } or
+    /**
+     * See `MatchRoute`
+     */
+    TMatchRoute(RouteBlock b, MethodCall m) { b.getAStmt() = m and m.getMethodName() = "match" }
+
+  /**
+   * A route configuration. This defines a combination of HTTP method and URL
+   * path which should be routed to a particular controller-action pair.
+   * This can arise from an explicit call to a routing method, for example:
+   * ```rb
+   * get "/photos", to: "photos#index"
+   * ```
+   * or via a convenience method like `resources`, which defines mutiple routes at once:
+   * ```rb
+   * resources :photos
+   * ```
+   */
+  abstract class Route extends TRoute {
```

Comment:
No `abstract` user-facing classes (unless the user is expected to contribute to the type).

Reply:
Hmm, but we need this class to be public to use it in `ActionControllerActionMethod::getARoute`. What's the best way to work around this? Maybe something like
```ql
class Route extends TRoute instanceof RouteRange {
  string getController { result = super.getController() }
  string getAction { result = super.getAction() }
}
abstract private class RouteRange extends TRoute { ... }

private class ExplicitRoute extends RouteRange, TExplicitRoute { ... }
...
```

Reply:
Yes, that should work. I tend to call such classes `RouteImpl`.